Add SearchStrategyRegistry and defaultSearchStrategy to support existing search behavior, and integrate it with CallClient.

cjcenizal opened the pull request:
Addresses initial steps in https://github.com/elastic/kibana/issues/20004:

* Create a registry for adding a handler that can create a search based on index pattern type.
* Refactor Courier to have a clear integration point for this registry.

## Handling errors

When you try to access a rollup index pattern in Discover without the correct search strategy (e.g. you've downgraded from X-Pack to OSS Kibana), you'll see an error:

![image](https://user-images.githubusercontent.com/1238659/42666529-90c55346-85fa-11e8-9072-c5b4f2249f0b.png)

This requires mentioning an X-Pack feature in OSS, but I talked about this with @AlonaNadler and we think it's acceptable since the user has already been using X-Pack to create the rollup originally.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6221/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6236/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6298/)

spalger commented on src/ui/public/courier/fetch/call_client.js:
```diff
@@ -106,28 +122,27 @@ export function CallClientProvider(Private, Promise, es) {
-        esPromise = es.msearch({ body: serializedFetchParams });
-        const clientResponse = await esPromise;
-        await respond(clientResponse.responses);
+        esPromise = Promise.all(esPromises);
```
Also, would you mind trying to write a test that validates that promises from the esClient are aborted correctly when a single request in a batch is aborted? I imagine it's going to be tricky but would probably be a good idea to have given the scope of this refactor.

cjcenizal replied:
> We'll need to update the esPromise logic to handle the esPromises array because it needs access to the individual esPromise.abort() methods

Can you see any problem with just removing the code that calls `esPromise.abort()`? What's the point of aborting them?

jen-huang commented on src/ui/public/courier/search_strategy/search_strategy_registry.js:
```diff
@@ -0,0 +1,69 @@
+
+  searchRequests.forEach(searchRequest => {
+    const matchingSearchStrategy = searchStrategies.find(searchStrategy => searchStrategy.isValidForSearchRequest(searchRequest));
+    const { id } = matchingSearchStrategy;
```
getting an error for a rollup index pattern in Discover, as it doesn't match any registered search strategies:

<img width="972" alt="screen shot 2018-07-06 at 3 56 25 pm" src="https://user-images.githubusercontent.com/1965714/42403322-36e8e8ac-8135-11e8-821e-f731c46360fc.png">

maybe for queries with no matching search strategies, we set id to `default`?

cjcenizal commented:
@jen-huang I ran into this error myself and then smacked my forehead with my palm. Technically, this situation won't ever arise, because you'll only be able to create a rollup index pattern using the UI in our `feature/rollups` branch, and that's where I'm also going to add the `rollup` specific strategy (WIP here: https://github.com/elastic/kibana/pull/20505). This branch is just supposed to create some abstractions we can use without changing any functionality.

That said, I wonder if we really need to gracefully handle cases where there's no matching strategy? That would mean that either a) a plugin developer relies on courier for searching but failed to import `'uiExports/search'` or b) we've created an index pattern which has a `type` defined that our UI doesn't currently support. Both of those seem like error cases to me, so I think we should probably handle them by throwing an informative error. WDYT?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6322/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6585/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6701/)

alexfrancoeur commented:
@cjcenizal @jen-huang I would hope that this doesn't happen that often and think an error here is just fine. Would this be shown when opening `rollup` visualizations as well? What happens if I open a dashboard that contains both `rollup` and `non-rollup` visualizations? Would we provide a warning or show only the raw data or would we show nothing? I imagine the latter given we can't use the rollup search endpoint. 

I'd love to have @gchaps input here for feedback on the wording. For instance rather than `Please use X-Pack to search this index pattern` it might make more sense to have something like `Please install an Elastic licensed version of Kibana to use rollups`. 

The alternative approach here is to hide any index patterns that were "rollup" index patterns. As a user, I think I'd prefer an error than thinking my data was deleted, but thought I'd throw it out there.

cjcenizal commented:
@alexfrancoeur 

>  Would this be shown when opening rollup visualizations as well?

Yes, I think it'd be best to treat rollup-based saved object errors in a consistent way.

> What happens if I open a dashboard that contains both rollup and non-rollup visualizations?

I think showing the warning would make the most sense, but I wonder what @elastic/kibana-sharing thinks.

Good idea about getting @gchaps's input here; that would be fantastic.

> The alternative approach here is to hide any index patterns that were "rollup" index patterns. As a user, I think I'd prefer an error than thinking my data was deleted, but thought I'd throw it out there.

Hiding saved objects which are based on rollups was also @AlonaNadler's suggestion. In a situation like that, you'd have to go into Management and review your index patterns to see that some of them are disabled because they're rollup based. I think this is a good idea too, but maybe somethings we should explore later on.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6814/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6818/)

cjcenizal commented:
@spalger @jen-huang This PR is ready for a review. It's passing the `callClient` tests I added, and I addressed Jen's concerns about handling cases where rollup support is missing, at least in Discover. I've updated the meta issue with to-do items for handling similar situations in visualizations and dashboards.

spalger commented on src/ui/public/courier/fetch/call_client.js:
```diff
@@ -34,43 +33,60 @@ export function CallClientProvider(Private, Promise, es) {
+          return ABORTED;
         }
+
+        const status = statuses[searchRequestIndex];
```
Pretty sure that `statuses` is not in the same order as `aggregatedSearchRequests`; `statuses` is sorted by original request order and `aggregatedSearchRequests` is sorted by strategy order, then original request order, so this isn't going to map the right strategies to the right requests. Should probably add some requests with different strategies to the `abort()` tests.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6908/)

cjcenizal commented on src/ui/public/courier/fetch/call_client.js:
```diff
@@ -34,43 +33,60 @@ export function CallClientProvider(Private, Promise, es) {
-
-            return responses[index];
+    const respondToSearchRequests = (responses = []) => {
+      const aggregatedSearchRequests = searchStrategiesWithRequests.reduce((allSearchRequests, { searchRequests }) => {
```
> then I realized it's probably because batching the requests by strategy can cause the requests/responses to be in a different order, so I suppose this is collecting the searchRequest in the order that the responses should be received. Is that right?

Right.

> I think instead of trying to match the ordering of aggregatedResponses we should instead try to pass responses to respondToSearchRequests in the original request order.

Great idea, thanks for the code.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6921/)

cjcenizal commented:
@spalger Ready for another look! I added tests and caught a bunch of mistakes.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6934/)

cjcenizal commented:
Retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/6939/)

cjcenizal commented:
Selenium:

```
02:43:14    │ proc  [ftr]        └- ✖ fail: "homepage app sample data dashboard should launch sample data set dashboard"
02:43:14    │ proc  [ftr]        │        tryForTime timeout: Error: tryForTime timeout: [POST http://localhost:9515/session/9acb461c6a21f33aae3da96ead70305e/element / {"using":"css selector","value":"[data-test-subj~=\"globalLoadingIndicator-hidden\"]"}] no such element: Unable to locate element: {"method":"css selector","selector":"[data-test-subj~="globalLoadingIndicator-hidden"]"}
```

X-Pack:

```
02:16:19  info  Installing from source
02:16:19    │ info  source path: /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/elasticsearch
02:16:19    │ info  install path: /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/kibana/.es/test-q1w09pw1lvg
02:16:19    │ info  license: trial
02:16:20    │ info  on master at 305bfea9c3c174d35136fddbf2bcd8544b4ea3bd
02:16:20    │ info  0 locally modified file(s)
02:16:20    │ info  ./gradlew :distribution:archives:tar:assemble
02:16:20    │ERROR  Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
02:16:51    │ERROR  Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
02:16:53    │ERROR  WARNING: An illegal reflective access operation has occurred
02:16:53    │ERROR  WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/var/lib/jenkins/.gradle/wrapper/dists/gradle-4.8.1-all/6fmj4nezasjg1b7kkmy10xgo2/gradle-4.8.1/lib/groovy-all-2.4.12.jar) to method java.lang.Object.finalize()
02:16:53    │ERROR  WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
02:16:53    │ERROR  WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
02:16:53    │ERROR  WARNING: All illegal access operations will be denied in a future release
02:17:00    │ERROR  warning: [options] bootstrap class path not set in conjunction with -source 8
02:17:01    │ERROR  1 warning
02:17:13    │ERROR  Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
02:17:13    │ERROR  WARNING: An illegal reflective access operation has occurred
02:17:13    │ERROR  WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/var/lib/jenkins/.gradle/wrapper/dists/gradle-4.8.1-all/6fmj4nezasjg1b7kkmy10xgo2/gradle-4.8.1/lib/groovy-all-2.4.12.jar) to method java.lang.Object.finalize()
02:17:13    │ERROR  WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
02:17:13    │ERROR  WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
02:17:13    │ERROR  WARNING: All illegal access operations will be denied in a future release
02:17:44    │ERROR  Note: /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/elasticsearch/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentGenerator.java uses or overrides a deprecated API.
02:17:44    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:19    │ERROR  Note: Some input files use or override a deprecated API.
02:18:19    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:19    │ERROR  Note: Some input files use unchecked or unsafe operations.
02:18:19    │ERROR  Note: Recompile with -Xlint:unchecked for details.
02:18:30    │ERROR  Note: Some input files use or override a deprecated API.
02:18:30    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:30    │ERROR  Note: /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/elasticsearch/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/JsonProcessor.java uses unchecked or unsafe operations.
02:18:30    │ERROR  Note: Recompile with -Xlint:unchecked for details.
02:18:31    │ERROR  Note: Some input files use or override a deprecated API.
02:18:31    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:31    │ERROR  Note: /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/elasticsearch/modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java uses or overrides a deprecated API.
02:18:31    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:34    │ERROR  Note: Some input files use or override a deprecated API.
02:18:34    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:35    │ERROR  Note: Some input files use or override a deprecated API.
02:18:35    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:37    │ERROR  Note: Some input files use or override a deprecated API.
02:18:37    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:42    │ERROR  Note: Some input files use or override a deprecated API.
02:18:42    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:42    │ERROR  Note: Some input files use unchecked or unsafe operations.
02:18:42    │ERROR  Note: Recompile with -Xlint:unchecked for details.
02:18:45    │ERROR  Note: /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/elasticsearch/x-pack/plugin/graph/src/main/java/org/elasticsearch/xpack/graph/action/TransportGraphExploreAction.java uses or overrides a deprecated API.
02:18:45    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:55    │ERROR  Note: Some input files use or override a deprecated API.
02:18:55    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:18:55    │ERROR  Note: Some input files use unchecked or unsafe operations.
02:18:55    │ERROR  Note: Recompile with -Xlint:unchecked for details.
02:19:00    │ERROR  Note: Some input files use or override a deprecated API.
02:19:00    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:19:01    │ERROR  Note: /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/elasticsearch/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportRollupSearchAction.java uses or overrides a deprecated API.
02:19:01    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:19:03    │ERROR  /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/elasticsearch/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/nio/SecurityNioHttpServerTransport.java:62: error: method sslConfiguration in class SSLService cannot be applied to given types;
02:19:03    │ERROR              this.sslConfiguration = sslService.sslConfiguration(SSLService.getHttpTransportSSLSettings(settings), Settings.EMPTY);
02:19:03    │ERROR                                                ^
02:19:03    │ERROR    required: Settings
02:19:03    │ERROR    found: Settings,Settings
02:19:03    │ERROR    reason: actual and formal argument lists differ in length
02:19:03    │ERROR  /var/lib/jenkins/workspace/elastic+kibana+pull-request+multijob-x-pack/elasticsearch/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/nio/SecurityNioHttpServerTransport.java:62: error: getHttpTransportSSLSettings(Settings) has private access in SSLService
02:19:03    │ERROR              this.sslConfiguration = sslService.sslConfiguration(SSLService.getHttpTransportSSLSettings(settings), Settings.EMPTY);
02:19:03    │ERROR                                                                            ^
02:19:04    │ERROR  Note: Some input files use or override a deprecated API.
02:19:04    │ERROR  Note: Recompile with -Xlint:deprecation for details.
02:19:04    │ERROR  Note: Some input files use unchecked or unsafe operations.
02:19:04    │ERROR  Note: Recompile with -Xlint:unchecked for details.
02:19:04    │ERROR  2 errors
02:19:04    │ERROR  
02:19:04    │ERROR  FAILURE: Build failed with an exception.
```

cjcenizal commented:
FYI I found a bug I need to diagnose address.

cjcenizal commented:
Couldn't repro the bug I found originally, so this PR is ready for review.

spalger commented on src/ui/public/courier/fetch/call_client.js:
```diff
@@ -79,65 +91,90 @@ export function CallClientProvider(Private, Promise, es) {
+        // Assigning searchRequests to strategies means that the responses come back in a different
+        // order than the original searchRequests. So we'll put them back in order so that we can
+        // use the order to associate each response with the original request.
+        const responsesInOriginalRequestOrder = new Array(searchRequestsAndStatuses.length);
```
Nothing "wrong" with this here, but I think it would be logical to use the `searchRequests` array instead of the `searchRequestsAndStatuses` array here, below, and in `respondToSearchRequests()`, don't you think? Especially since `respondToSearchRequests()` is reading the status from `searchRequestsAndStatuses` by index anyway.

cjcenizal replied:
At least for me, using `searchRequestsAndStatuses` makes the code a bit easier to follow by reducing the number of variables I have to keep in mind. In terms of data structures being consumed, `searchRequests` stops being important immediately, and `searchRequestsAndStatuses`, `requestsToFetch`, and `searchStrategiesWithRequests` become the important ones. It would be a bit confusing to me if I were to read this code and see `searchRequests` was reintroduced -- I'd wonder what the intention behind that was.

spalger commented on src/core_plugins/kibana/public/discover/directives/unsupported_rollup.js:
```diff
@@ -0,0 +1,43 @@
+    <EuiFlexGroup justifyContent="center">
+      <EuiFlexItem grow={false} className="discoverNoResults">
+        <EuiCallOut
+          title="OSS Kibana doesn't support rollup index patterns. Please use X-Pack to search this index pattern."
```
Do you think users will understand "OSS Kibana"? I wonder if we should rephrase this to instead say something like: "Index Patterns based on rollup indices require the rollup plugin from X-Pack which is either not installed or disabled."

cjcenizal replied:
I like this a lot. @gchaps thoughts?

spalger commented on src/ui/public/courier/fetch/call_client.js:
```diff
@@ -79,65 +91,90 @@ export function CallClientProvider(Private, Promise, es) {
+        // order than the original searchRequests. So we'll put them back in order so that we can
+        // use the order to associate each response with the original request.
+        const responsesInOriginalRequestOrder = new Array(searchRequestsAndStatuses.length);
+        segregatedResponses.forEach((responses, strategyIndex) => {
```
Maybe replace `responses` on this line with `{ responses }` to avoid some of the redundancy here?

cjcenizal replied:
Maybe I should just change the interface to only return the `responses` array. WDYT?

spalger replied:
Either works for me

jen-huang commented on src/ui/public/courier/search_strategy/search_strategy_registry.js:
```diff
@@ -0,0 +1,92 @@
+  return Boolean(getSearchStrategy(indexPattern));
+};
+
+const isRollupIndexPattern = indexPattern => {
```
I don't think we should have any rollup references in OSS code. maybe this can be abstracted out as part of as part of registered search strategies? see my next comments for ideas 😄

jen-huang commented on src/core_plugins/kibana/public/discover/controllers/discover.js:
```diff
@@ -777,5 +777,16 @@ function discoverController(
 
+  // Block the UI from loading if the user has loaded a rollup index pattern but it isn't
+  // supported.
+  $scope.isUnsupportedRollup = (
```
when rollup reference is abstracted away, maybe this can be something like: 

```js
import {
    hasSearchStategyForIndexPattern,
    isIndexPatternSupported,
    renderUnsupportedIndexPatternMessage,
} from 'ui/courier';

if(
    !hasSearchStategyForIndexPattern($route.current.locals.ip.loaded)
    || !isIndexPatternSupported($route.current.locals.ip.loaded)
) {
   // psuedo code - implementation will need adjustment for rendering logic! :)
    $scope.unsupportedMessage = renderUnsupportedIndexPatternMessage();
    return;
}

jen-huang replied:
I redact my previous review after chatting with CJ and realizing that the unsupported rollup messages were are due to giving user hints when they're using OSS. X-Pack code isn't reachable in that scenario.

The comprise to support this type of error in OSS, while avoiding cross stream OSS/X-Pack code and rollup references, is to check whether the index pattern is the default type, instead of checking for rollup type.

jen-huang commented on src/core_plugins/kibana/public/discover/directives/index.js:
```diff
@@ -25,8 +25,14 @@ import {
   DiscoverNoResults,
 } from './no_results';
 
+import {
```
references to rollup message can be removed with implementation of `renderUnsupportedIndexPatternMessage` or similar

jen-huang commented on src/core_plugins/kibana/public/discover/directives/unsupported_rollup.js:
```diff
@@ -0,0 +1,43 @@
+/*
```
then, this component can be moved somewhere inside x-pack/rollups directory

jen-huang commented on src/core_plugins/kibana/public/discover/index.html:
```diff
@@ -60,6 +60,10 @@ <h1 tabindex="0" id="kui_local_breadcrumb" class="kuiLocalBreadcrumb">
 
       <div class="discover-wrapper col-md-10">
         <div class="discover-content">
+          <discover-unsupported-rollup
```
and this will use generic `$scope.unsupportedMessage` or similar, instead of referencing rollups directly

cjcenizal commented:
@spalger @jen-huang I've updated the error message and removed references in the code to rollups (except for comments). I've also changed `callClient` to expect a strategy's `searching` promise to resolve with a `responses` array.

![image](https://user-images.githubusercontent.com/1238659/42787159-22e1f02a-890e-11e8-8992-6aa64e755209.png)

spalger commented:
Still LGTM

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7018/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7024/)

spalger commented on src/ui/public/courier/search_strategy/default_search_strategy.js:
```diff
@@ -68,12 +68,12 @@ export const defaultSearchStrategy = {
 
-    const searching = es.msearch({ body: serializedFetchParams }).then(({ responses }) => responses);
-    const abort = searching.abort;
+    const searching = es.msearch({ body: serializedFetchParams });
```
👍

gchaps commented:
@cjcenizal A minor edit.  Add a comma before which and I think we can remove "either"

Index patterns based on rollup indices require the rollup plugin from X-pack, which is not installed or disabled

gchaps commented:
@cjcenizal @alexfrancoeur 

re: the wording of  `Please use X-Pack to search this index pattern` or `Please install an Elastic licensed version of Kibana to use rollups.`

How about:

"OSS Kibana doesn't support rollup index patterns. You must use the default distribution of Kibana and a Basic license at a minimum."

I picked "default distribution" due to the wording here: [https://www.elastic.co/downloads/kibana](https://www.elastic.co/downloads/kibana)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7130/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/7137/)
